src/app/loadout-builder/mod-assignments.ts

```diff
@@ -0,0 +1,140 @@
+import { DestinyEnergyType } from 'bungie-api-ts/destiny2';
+import _ from 'lodash';
+import { DimItem, PluggableInventoryItemDefinition } from '../inventory/item-types';
+import { ProcessItem, ProcessMod } from './process-worker/types';
+
+const DEFAULT_ENERGY_INVESTMENT = 1000;
+
+interface ItemAssignments {
+  mods: ProcessMod[];
+  energySwapped: boolean;
+}
+
+export class ModAssignments {
+  // We just keep track of the id's and hashes as we assume only valid items and mod
+  // combninations are passed in
+  private slotDependantAssignments: Map<string, number[]>;
+  // We need to keep track of a bit more data here so we track the item and the data
+  // kept in ItemAssignments
+  private slotIndependantAssignments: Map<ProcessItem, ItemAssignments>;
+  // Keep track of the current set of assignments extra energy investment so we dont
+  // need to continually recalculate it
+  private extraEnergyInvestment = DEFAULT_ENERGY_INVESTMENT;
+
+  constructor() {
+    this.slotDependantAssignments = new Map();
+    this.slotIndependantAssignments = new Map();
+  }
+
+  get isSuccessfullyAssigned() {
+    return this.extraEnergyInvestment < DEFAULT_ENERGY_INVESTMENT;
+  }
+
+  /**
+   * This returns a map of item id's to mod hashes for the assigned mods.
+   */
+  getResults() {
+    const rtn = new Map<string, number[]>();
+
+    for (const [item, itemAssignments] of this.slotIndependantAssignments.entries()) {
+      rtn.set(
+        item.id,
+        itemAssignments.mods.map((mod) => mod.hash)
+      );
+    }
+
+    for (const [itemId, mods] of this.slotDependantAssignments.entries()) {
+      if (rtn.has(itemId)) {
+        rtn.get(itemId)?.push(...mods);
+      } else {
+        rtn.set(itemId, mods);
+      }
+    }
+
+    return rtn;
+  }
+
+  /**
+   * Assigns slot dependant mods (i.e. helmet mods) to the item.
+   */
+  assignSlotDependantMods(item: DimItem, mods: PluggableInventoryItemDefinition[]) {
```

Comment:
From an API design standpoint I don't like these kinds of mutable classes - it's hard for me to understand how they're supposed to be used. What's the state of the object when it's been freshly instantiated? What's its state after `assignSlotDependentMods`? How about `assignSlotIndependantModsIfLessEnergyTypeSwaps`?

I'd try really hard to figure out how to just make all of this a function or set of functions that returns a result.

Reply:
Slot dependant mods would force a energy type change so setting them just happens, there is no special handling of trying to work out a minimum energy change.

Regarding the API design and this mutable class, the logic is pretty hairy to figure this out so I was trying to segregate it to it's own class. I could potentially change it to a completely functional model but we were already mutating stuff to keep track of this. I will need to come up with some way to keep track of the current best and compare it to a new set in a performant manner.